File: app-laravel/src/Product/Application/Queries/ListProducts/ListProductsHandler.php
<?php

declare(strict_types=1);

namespace Src\Product\Application\Queries\ListProducts;

use Src\Product\Application\DTOs\ProductDTO;
use Src\Product\Application\Queries\ListProducts\ListProductsQuery;
use Src\Product\Infrastructure\Models\ProductModel;
use Src\Shared\Application\Bus\Query;
use Src\Shared\Application\Bus\QueryHandler;

/**
 * ListProductsHandler — Filtrlənmiş və səhifələnmiş məhsul siyahısı.
 *
 * AXIN:
 * 1. Query-dən filter parametrlərini al
 * 2. ProductFilter::apply() ilə query-yə filterləri tətbiq et
 * 3. Paginate et
 * 4. Hər məhsulu ProductDTO-ya çevir
 */
final class ListProductsHandler implements QueryHandler
{
    public function handle(ListProductsQuery $query): mixed
    {
        // Eloquent query builder başlat
        $eloquentQuery = ProductModel::query();

        // Filterləri tətbiq et (search, price range, stock, sort)
        ProductFilter::apply($eloquentQuery, $query->filters());

        // Səhifələ
        $paginated = $eloquentQuery->paginate(
            perPage: $query->perPage(),
            page: $query->page(),
        );

        // DTO-ya çevir
        $paginated->getCollection()->transform(
            fn($product) => ProductDTO::fromModel($product),
        );

        return $paginated;
    }
}

File: app/spring/src/main/java/az/ecommerce/product/application/query/listproducts/ListProductsHandler.java
package az.ecommerce.product.application.query.listproducts;

import az.ecommerce.product.application.dto.ProductDto;
import az.ecommerce.product.domain.repository.ProductRepository;
import az.ecommerce.shared.application.bus.QueryHandler;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class ListProductsHandler implements QueryHandler<ListProductsQuery, List<ProductDto>> {

    private final ProductRepository repository;

    public ListProductsHandler(ProductRepository repository) { this.repository = repository; }

    @Override
    public List<ProductDto> handle(ListProductsQuery q) {
        return repository.findAll(q.page(), q.size()).stream().map(ProductDto::fromDomain).toList();
    }
}
